support custom handlebars thru plugin config

isao opened the pull request:
make locator-handlebars aware of config for optional alternate Handlebars instance to use for server-side use. See also https://trello.com/c/PziJXjiL

todo
- [x] docs
- [x] tests

Commit:
initial attempt to allow user-provided handlebars

- pass a user-provided hb instance via plugin like
    .plug(LocatorHandlebars.yui({handlebars: handlebars}))
- allow that to override the locally installed hb

Commit:
try app.yui.applyGroupConfig

Commit:
fix core.js tests

caridy commented:
LGTM

caridy commented on example/app-userhb.js:
```diff
@@ -0,0 +1,70 @@
+/*jslint node:true, nomen: true*/
```
Probably it is time to split `example` into `examples`, having more that one example is what we do in other components.

isao replied:
ok, will do.

Commit:
add test for default & custom hb getting applied; whsp

Commit:
add tests for HB extensions config

Commit:
add HB & extensions as config options

Commit:
rm custom HB example

isao commented:
@caridy I added some tests and updated the readme to include info on customizing the HB instance and HB's extensions. I also removed the `example/app-userhb.js`. Are these points below ok with you?
- I don't think it's worthwhile to make a second example app showing the custom HB, since the client-side part is out-of-scope ATM, and the diff is only a few LOC.
- I left out describing other things that can be added to the `LocatorHandlebars.yui(options)` config in the README, since it might be confusing (and are unrelated to this PR).

caridy commented on example/package.json:
```diff
@@ -7,6 +7,7 @@
     "dependencies": {
         "express": "*",
         "express-yui": "*",
+        "handlebars": "*",
```
not needed anymore.

isao replied:
ah, thanks for the catch. removed.

caridy commented on lib/plugin.js:
```diff
@@ -8,20 +8,22 @@
 
-var core = require('./core');
+var core = require('./core'),
+    HB = require('yui/handlebars').Handlebars;
```
@isao, I remember that @ericf wanted to use `require('handlebars')` directly, instead of the version shipped with YUI. Apparently, the issue with the error when using a different version is not longer a case since YUI is not modifying handlebars in any way starting in 3.12.x, I might be wrong here.

isao replied:
yes, I remember that comment. there are a few [minor differences](http://yuilibrary.com/yui/docs/handlebars/#differences-from-original-handlebars), but keeping `yui/handlebars` for now is also a way to ensure that the server-side and client-side versions are same by default.

caridy commented on example/package.json:
```diff
@@ -7,6 +7,7 @@
+        "handlebars": "*",
         "locator": "*",
         "locator-handlebars": "*",
         "yui": "~3.11.0"
```
Part of the deal here was to relax this dependency as well, or removing it completely.

caridy commented on tests/lib/index.js:
```diff
@@ -35,6 +35,44 @@ describe('locator-handlebars', function () {
+        it('YUI/handlebars applied to index.yui().handlebars by default', function () {
+            var extended = index.yui(),
+
+                expected = '1.0.0-yui',
```
where does this value comes from? it seems like a flaky tests.

isao replied:
the actual assertion only considers the last four characters, which for the YUI version, should always be "-yui".

I formatted it this way to display a realistic (actual, for now) value. I can change it to:

```
expected = '1.0.0-yui'.slice(-4),
actual = extended.describe.handlebars.VERSION.slice(-4);
```

but I'm not sure that reads any better...

caridy replied:
Yeah, but what's the point of this test? it only tests that yui brings handlebars, but I guess the methods are more important that a random version that is dev time defined. Just check if that thing is an object and has at least one of the basic methods, that will never change. What if @ericf decides to modify that VERSION and use the original version number without the yui suffix? it will break our test.

isao replied:
I was just testing that, by default, the handlebars object was the YUI one. True, if the suffix changes the test will break, though the "-yui" suffix is officially documented.

I'll update it.

isao replied:
updated.. comparing actual objects now, which is better. Thanks @caridy

Commit:
rm outdated example app dependency

Commit:
change '-yui' version-substring assertion to object comparison

caridy commented:
+1 lets merge this and relax the dependencies later on.

caridy commented:
@isao are you going to release the new version now that this is merged?

isao commented:
@caridy ok tagged & released 0.2.2 on github, but I need to be added to the npm owners to publish..

```
Error: forbidden user: isao not authorized to modify locator-handlebars
```

caridy commented:
added, you can publish it now!

isao commented:
Thanks!

isao commented:
published

% npm info --log error locator-handlebars dist-tags
{ latest: '0.2.2' }
